Add IPv4/IPv6 dual stack KEP

leblancd opened the pull request:
Add a KEP for IPv4/IPv6 dual stack functionality to Kubernetes clusters. Dual-stack functionality includes the following concepts:
- Awareness of multiple IPv4/IPv6 address assignments per pod
- Native IPv4-to-IPv4 in parallel with IPv6-to-IPv6 communications to, from, and within a cluster

References:
[kubernetes/kubernetes issue # 62822](https://github.com/kubernetes/kubernetes/issues/62822)
[kubernetes/features issue # 563](https://github.com/kubernetes/features/issues/563)

feiskyer commented:
/cc @kubernetes/sig-network-proposals

rpothier commented:
/area ipv6

caseydavenport commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+- Link Local Addresses (LLAs) on a pod will remain implicit (Kubernetes will not display nor track these addresses).
+- Kubernetes needs to be configurable for up to two service CIDRs.
+- Backend pods for a service can be dual stack. For the first release of dual-stack support, each IPv4/IPv6 address of a backend pod will be treated as a separate Kubernetes endpoint.
+- Kube-proxy needs to support IPv4 and IPv6 services in parallel (e.g. drive iptables and ip6tables in parallel).
```
We should also consider the impact to the IPVS proxy. I think at this point we need to maintain both (unless we say dual-stack is iptables only?)

leblancd replied:
Thanks for looking into this, it's good to know up front!

caseydavenport commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+### Awareness of Multiple IPs per Pod
+
+Since Kubernetes Version 1.9, Kubernetes users have had the capability to use dual-stack-capable CNI network plugins (e.g. Bridge + Host Local, Calico, etc.), using the 
+[0.3.1 version of the CNI Networking Plugin API](https://github.com/containernetworking/cni/blob/spec-v0.3.1/SPEC.md), to configure multiple IPv4/IPv6 addresses on pods. However, Kubernetes currently captures and uses only the first IP address in the list of assigned pod IPs that a CNI plugin returns to Kubelet in the [CNI Results structure](https://github.com/containernetworking/cni/blob/spec-v0 ...
```
I believe Kubernetes uses only the IP address it reads from eth0 within the Pod and ignores the response from CNI altogether, right? Or did that get changed and I missed it? :)

squeed replied:
That is still currently the case.. for now. CNI get is getting closer!

squeed replied:
CNI doesn't allow for any kind of annotations (right now -- that could change) - it only has IPs and routes. Changing that is... out of scope :-)

I think this section is fine as-is; how kubelet gets the list of IPs from a running container is just an implementation detail.

leblancd replied:
@squeed - Understood, thanks. This was more out of curiosity about where we're headed with the CNI API.

caseydavenport commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+
+### Support of Health/Liveness/Readiness Probes
+
+Currently, health, liveness, and readiness probes are defined without any concern for IP addresses or families. For the first release of dual-stack support, no configuration "knobs" will be added for probe definitions. A probe for a dual-stack pod will be deemed successful if either an IPv4 or IPv6 response is received. (QUESTION: Does the current probe implementation include DNS lookups, or are  ...
```
> Does the current probe implementation include DNS lookups, or are IP addresses hard coded

I believe you can configure a `host` for the probe which will be resolved via DNS if it is a DNS name rather than an exact IP. [See here](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#configure-probes)

thockin replied:
Specifically this means that probes will execute against the primary IP only ?

caseydavenport commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+
+### Load Balancer Operation ???
+
+### Network Policy Considerations ???
```
Speaking strictly for Calico, I don't believe there are any (besides testing). 

I don't think there are API impacts because the NP API selects using labels rather than addresses (though now that I think about it we should check if the NP CIDR support does validation on IPv4 / IPv6)

danwinship commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+
+#### First Release of Dual Stack: No Service Address Configuration "Knobs"
+For the first release of Kubernetes dual-stack support, no new configuration "knobs" will be added for service definitions. This greatly simplifies the design and implementation, but requires imposing the following behavior:
+- Service IP allocation: Kubernetes will always allocate a service IP from each service CIDR for each service that is created. (In the future, we might want to consider adding configuration options to allow a user to select e.g. whether a given service should be assigned only IPv4, only IPv6, or both IPv4 and IPv6 service IPs.)
```
If kube-proxy/plugins/etc are going to have to be updated to deal with multiple CIDR ranges anyway, then it would be useful to let the admin say "a.b.c.d/x should be treated as part of the service IP range, but the controller shouldn't allocate any new IPs out of that range". This would let you live-migrate the cluster from one service CIDR to another (something we (OpenShift) have had a handful of requests for).

(The same applies to the cluster CIDRs, though in that case the allocations are done by the plugin, not kube itself, so you could already implement this via plugin-specific configuration.)

thockin replied:
I have heard a VERY small number of similar requests, but I think it's orthogonal to multi-family.  We could add multi-CIDR with metadata (e.g. allow existing, but no more allocations) without touching multi-family support.

That said, *if* we do add multiple CIDRs, we should at least make it possible to add metadata later (so []struct, not []string).

squeed commented:
A few thoughts about ExtraPodIPs:
- Should it always also include the default PodIP? So later clients don't have to manually concatenate
- I would like some metadata attached to IPs. Some labels, perhaps?

Some time ago, before multi-network was tabled, I suggested that every IP should be labeled with the name of the network that created it. Then services could have an optional label selector.

Even if we don't use this right now, it would be good to have for room to grow.

squeed commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+The singular --service-cluster-ip-range argument will become deprecated.
+
+#### controller-manager Startup Configuration for Multiple Service CIDRs
+A new, plural "service-cluster-ip-ranges" option for the [controller-manager startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/) is proposed, in addition to retaining the existing, singular "service-cluster-ip-range" option (for backwards compaibility):
```
complete utter bikeshed: what about making `--service-cluster-ip-range` accept a comma-separated list?

thockin replied:
if we can do that transparently, that might be great for all of these flags (if a bit less obviously named)

squeed commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
````diff
@@ -0,0 +1,362 @@
+A new, plural "service-cluster-ip-ranges" option for the [controller-manager startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/) is proposed, in addition to retaining the existing, singular "service-cluster-ip-range" option (for backwards compaibility):
+```
+  --service-cluster-ip-range  ipNet         [Singular IP CIDR,  Default: 10.0.0.0/24]
+  --service-cluster-ip-ranges stringSlice   [Multiple IP CIDRs, comma separated list of CIDRs, Default: []]
````
what happens if the user specifies two v4 ranges?

thockin replied:
good call - err on the side of over-specifying :)

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+This proposal aims to extend the Kubernetes Pod Status API so that Kubernetes can track and make use of multiple IPv4/IPv6 address assignments per pod.
+
+#### Versioned API Change: PodStatus v1 core
+In order to maintain backwards compatibility for the core V1 API, this proposal adds a new "ExtraPodIPs" field to the core V1 version of the [PodStatus V1 core API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#podstatus-v1-core), while maintaining the existing "PodIP" field. This change is in accordance with the [Kubernetes API change
```
for context, maybe include an appendix or footnote detailing this?

The issue with compat here is older clients.

First: We can't really stop populating this field, or we break clients who don't understand `PodIPs`.

A simpler-seeming answer might be to have both `PodIP` and `PodIPs` where the plural always includes the singular.  Again, the problem is older clients (think about a failed kubelet update which gets rolled back).

Consider an existing pod with `PodIP` = "X" and `PodIPs" = [ "X", "Y" ].  If we have a client that only knows `PodIP`, and it updates that field to "Z", but does not update `PodIPs`, the inclusivity invariant is violated.

If we make it optionally inclusive, we still may have muddled the intention - the client changed "X" to "Z", did they mean that "X" is no longer preferred or no longer present at all?

The same problem exists in the other direction. If the older client updated `PodIP` from "Z" to "X", that collides with `PodIPs`.  Did the user mean the same "X"?

That doesn't leave much room for viable solutions.  It can't be guaranteed exclusive nor can it be guaranteed inclusive, and optionally inclusive is confusing at best.  The only real option here is to consider the real-life chance of detriment.  Pod IPs *do* change (e.g. if the node reboots quickly, it can come back without evicting the pod, but IPAM can choose a new IP), but it's not expected to be common.  It's also uncommon for multiple clients to be writing to Pod.status, so we really only need to consider the upgrade/rollback case for mutations.

Thus, I think it's reasonably safe to say it is optionally inclusive, and to define the edge case behavior.

This might be worth having a very focused sig-architecture chat about.  There may be better options that I haven't thought about or precedents that I don't know.

Other options, off the top of my head:

* Linked fields.  Any update to singular with a corresponding update to plural can validate inclusivity/exclusivity.  Any update to singular without an update to plural forces plural to some defined state (inclusive or exclusive TBD).  Any update to plural without an update to singular forces singular to some defined state.

* move Pod to v2, make plural be read-only in v1 and collaps singular and plural in v2.  Changes through v2 extract the first item (or some other heuristic) into singular in v1 view.

danwinship replied:
> There may be better options that I haven't thought about or precedents that I don't know.

The "[API changes](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#on-compatibility)" doc explicitly recommends this approach. (The example with `Param` and `ExtraParams`.)

squeed replied:
Ah, that makes sense. I hadn't thought about kubelet downgrades (scary!).

I would really like to attach metadata to IPs - thoughts about how?

thockin replied:
I think we need it too.  This is a pattern (pluralising a field) that doesn't pop up often enough to have a pattern that I know of, so maybe we will have to trailblaze a bit.

pmichali replied:
Are there two different suggestions here? One being that the old field has the first (primary) value and the array has additional entries (seems like the API changes doc specifies this). The other being the new array containing the entry held in the original field (duplicating)? Just trying to get this straight.

Would it be good to have a separate section to discuss field extension in general, the pros, cons, corner cases, etc., since we have several of these?

leblancd replied:
@pmichali - Right, the two *main* options being discussed could be called exclusive, where you have 
```
podIP="X"
podIPs=["Y", "Z"]
```
versus inclusive, where you have:
```
podIP="X"
podIPs=["X", "Y", "Z"]
```
The exclusive is what was originally proposed, and is based on the pluralized example in the [API changes doc](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#on-compatibility). And then there's an "optional inclusive" variant that seems to be a inclusive/exclusive hybrid (old clients only update the singular podIP, so podIPs inclusivity is lost upon modification of podIP). I had been thinking that either pure exclusive or pure inclusive models should work - old clients remain agnostic of the pluralized field, but the conversion logic in the server (stuff that converts from versioned APIs to internal versions, and vice versa) is smart enough to maintain coherency in the podIP and podIPs fields... but I could be missing nuances.

Yes, the "extra section" is what @thockin is suggesting (in appendix or footnote form), although I'm thinking that if the  [API changes doc](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#on-compatibility) can be updated accordingly, we could just point to that.

thockin replied:
@leblancd Not everyone will do a read-modify-write, and even if they do ... older clients will drop unknown fields AND they can send a patch.  So pathologically:

Mandatory exclusive:
* podIP="X", podIPs=["Y", "Z"]
* client sends a patch operation changing podIP to "Z".
=> If we fail that at the apiserver, the client has no idea why an update that used to work doesn't any more.  If we don't fail it, we can either remove "Z" from podIPs (mutating user input is generally verboten) or we can ignore it and break the invariant.

Mandatory inclusive:
* podIP="X", podIPs=["X", "Y"]
* client sends a patch operation changing podIP to "Z".
=> If we fail that at the apiserver, the client has no idea why an update that used to work doesn't any more.  If we don't fail it, we can either add  "Z" to podIPs (mutating user input is generally verboten) and remove "X" (loses user metadata) or we can ignore it and break the invariant.

I'm saying we need to find a cleaner way and set precedent very carefully.  Optionally-inclusive (which I guess is == optionally exclusive :) is still confusing.
* podIP="X", podIPs=["X", "Y"]
* client sends a patch operation changing podIP to "Z".
=> Is the "X" in podIPs valid any more?

I want to consider more clever ideas (if anyone has some).  This *will* happen again.

leblancd replied:
@thockin It's not clear to me how we can solve this.

We talked elsewhere in this review about including metadata (e.g. annotations) along with IPs in the podIPs/extraPodIPs array. If we do include metadata, I think we'd be better off (or it would at least be simpler) using an inclusive model. The reason being that the singular podIP field needs to remain a simple IP address (the "default" IP address) without metadata. If we use an inclusive model, we can store metadata associated with the default IP address in the podIPs array. (For exclusive, we'd have to add yet another field for that IP address' metadata).

Of course, if an old client changes podIP, the old client is agnostic to metadata, so the new podIP will be added without metadata. The podIPs field would then have a mix of entries with/without metadata. I don't know how bad of a problem this would be, but there's probably no way around it.

For the mandatory inclusive scenario, I would consider the "X" in podIP and the "X" in podIPs to be duplicate copies of the same field, so I would think that if client replaces/overwrites the "X" in podIP with "Z", then "Z" should be added to podIPs, and the (duplicate) "X" in podIPs should be removed, along with its metadata, IMO. Is mutating user input so bad that we shouldn't even consider it?

I'm looking for guidance. In the meantime, I'll rewrite this section assuming a mandatory inclusive model for now, because of the metadata concerns I mentioned above.

thockin replied:
@leblancd I think it is about harm reduction.  If someone uses an older client, it can't stop working.  In practice I doubt we will have many problems with multi-writer, so I think we can opt for the optionally-inclusive mode.

I've put this on the agenda for sig-arch this Thursday, though we have a pretty full list, so it may push to next week.

https://docs.google.com/document/d/1BlmHq5uPyBUDlppYqAAzslVbAO8hilgjqZUTaNXUhKM/edit?ts=5b43a34a#

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+    // Additional IP addresses allocated to the pod. Routable at least within
+    // the cluster. Empty if not yet allocated.
+    // +optional
+    ExtraPodIPs []string `json:"podIP,omitempty" protobuf:"bytes,6,opt,name=ExtrapodIPs"`
```
and I am not sure map[string]string is what we want either.  Worth thinking about what other data we might want here, or maybe even just using a struct with only one field for future expansion?

lavalamp replied:
It does seem like naming them might be useful, so you can refer to them before they have been assigned.

pmichali commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+#### Kube-Proxy Startup Configuration Changes
+
+##### Multiple bind addresses configuration
+New [kube-proxy configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) arguments will be added to allow a user to specify separate IPv4 and IPv6 bind addresses for the kube-proxy server:
```
Mentioned above, would it suffice to just support health checks for one IP mode (IPv6, if available, else IPv4)?

For metrics, wouldn't only one IP mode be needed to access the metrics? Not sure why one would want to access metrics from IPv4 and IPv6. May be able to leave that, as is.

pmichali commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+## Drawbacks [optional]
+\<TBD\>
+
+## Alternatives [optional]
```
I think this would be good to have up top, where we discuss the motivation, and then the ways to address the needs and the selection chosen.

pmichali commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+## Implementation History
+\<TBD\>
+
+## Drawbacks [optional]
```
Would be good to list the limitations/restrictions of the proposal as one of the prominent items, IMHO. Would cover this and the Risks/Mitigations item).

pmichali commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+- Kube-dns is capable of running dual-stack, but it is currently only made aware of either all IPv4 or all IPv6 service addresses.
+
+For scenarios that require legacy IPv4-only clients or services, the above restrictions mean that complex and expensive IPv4/IPv6 transition mechanisms (e.g. NAT64/DNS64, stateless NAT46, or SIIT/MAP) will need to be implemented in the data center networking.
+
```
It would be useful to have a section on the use cases that need to be addressed, IMHO, so that people have a picture of how the proposal(s) would address the needs. I know, it's a big ask...

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
````diff
@@ -0,0 +1,362 @@
+  --pod-cidr   string        [Singular IP CIDR,  Default: ""]
+  --pod-cidrs  stringSlice   [Multiple IP CIDRs, comma separated list of CIDRs, Default: []]
+```
+The singular --pod-cidr argument will become deprecated.
````
will these flags be mutually exclusive?  Seems like they should be.  what happens if both are specified (hint: older has to win)

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
````diff
@@ -0,0 +1,362 @@
+  --cluster-cidr   string        [Singular IP CIDR,  Default: ""]
+  --cluster-cidrs  stringSlice   [Multiple IP CIDRs, comma separated list of CIDRs, Default: []]
+```
+The singular --cluster-cidr argument will become deprecated.
````
same as above

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+
+### Support for Multiple Service IPs per Service
+
+This proposal adds the capability to define multiple service CIDRs for a cluster.
```
Do we HAVE TO do this?  As a simplifying assumption, what if we said that every cluster has a "primary" address family.  All pods have an address in that family and all Services addresses come from that family.  Pods can have "extra" IPs in other families...

On one hand this seems to simplify Service and Endpoint logic.  On the other hand I am honestly asking "can we get away with this" ?  I don't know.

Or maybe we can get away with this assumption for now, and revisit this later?  The tie-in with multi-net and multi-tenant here is obvious and those topics are not settled, IMO.

Thoughts?

thockin replied:
Also we should consider how to implement this.  The IP allocation logic depends on etcd, so allocating from a static number of CIDRs should be relatively easy.  Allocating from N will be more complex, and we have the same compat / upgrade / rollback issues to deal with but at a less visible layer of the stack.

leblancd replied:
That is a VERY good question. I've thought about what we would be giving up with this approach. From a high level, I think what we'd end up with is, roughly speaking (assuming an IPv6 service CIDR):
- Internal: V6-only support inside the cluster (no support for legacy IPv4-only pods)
- External: Dual-stack external access to/from the cluster (excluding external load balancers, maybe some ingress controllers?)

More detail (please correct any wrong assumptions that I might have made here):

- IPv4-only pods: IPv4-only client pods couldn't be supported because they could not do service discovery. If these pods do a DNS lookup for a Kubernetes service, they would only care about A records. So we'd have to have an IPv4 address that maps somehow to an IPv6 service IP. We could then potentially do a 4-to-6 mapping of packets, but we'd have to figure out where that mapping gets done (add a sidecar to the pod?). It would be very messy and complicated, at best. Similarly, an IPv4-only backend pod couldn't be supported, unless very complicated DNS address and packet mappings were added.
- Ingress: I believe that this would provide dual-stack access from the outside, even if there in only on service CIDR, at least for an Nginx ingress controller. The Nginx ingress controller should be able to support IPv4 and IPv6 connections from the outside, and load balance them to both IPv4 and IPv6 pod endpoints. (I'm assuming that Kubernetes will still create both IPv4 and IPv6 endpoints, even in the absence of any IPv4 service IPs).
- External Load Balancers: I believe that dual-stack access won't work for at least some external load balancers because they map to a service IP, and let kube-proxy do the load balancing?
- Exporting a Service via NodePort: This could *possibly* provide dual stack support even if there is only a single family service CIDR, BUT... we would have to still support both iptables and ip6tables in parallel, since both tables would need to be listening on that NodePort.
- Exporting a Service via ExternalIP: Same as NodePort, this could *possibly* provide dual stack support even if there is only a single family service CIDR, BUT... we would have to support both iptables and ip6tables in parallel, since iptables would need to be listening for IPv4 ExternalIPs, and ip6tables would need to be listening for IPv6 ExternalIPs.

thockin replied:
So, I don't have an answer - that's why I was asking. :)

I think all of your points are correct.  It's a little bit of mental gymnastics to think about, but it is simpler overall.  Examples:

* If your "external" LBs support v6, you probably should make v6 your primary (any reason not to?).  Pods can optionally have v4 addresses (why would someone do this?)

* If your "externa;" LBs only support v4, you should make v4 your primary, and some or all pods can have v6 addresses (e.g. intra-corp traffic).

Do we have any real use-cases where someone would want some pods to be dual-stack and some not be?  You have more v6 customers than me :)

leblancd replied:
I'm rewriting this spec to propose using just a single service CIDR (with a section at the end in "Alternatives" discussing the tradeoffs compared to dual service CIDR). I'm thinking that the dual (or more generally, multiple) service CIDR support may take more work than it's worth. I don't know of any specific cases where customers require e.g. an IPv4-only app running in an IPv6-only cluster. As you say, anyone who has IPv4-only (IPv6-only) apps can run their cluster using an IPv4 (IPv6) service CIDR. Plus, in the unlikely event that we find that we *really* need dual service CIDRs, we can add it (or a clever workaround) in a later release.

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+(Note that this assumed behavior works very well for the typical dual-stack case where there are one IPv4 and one IPv6 service CIDR, and one IPv4 and one IPv6 address per pod.)
+
+#### Versioned API Change: Service v1 core
+In order to maintain backwards compatibility for the core V1 API, this proposal adds a new "ExtraClusterIPs" field to the core V1 version of the [Service V1 core API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#service-v1-core), while maintaining the existing "ClusterIP" field. This change is in accordance with the [Kubernetes API change
```
same discussion as above :)

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+The [Kubernetes ingress feature](https://kubernetes.io/docs/concepts/services-networking/ingress/) relies on the use of an ingress controller. The two "reference" ingress controllers that are considered here are the [GCE ingress controller](https://github.com/kubernetes/ingress-gce/blob/master/README.md#glbc) and the [NGINX ingress controller](https://github.com/kubernetes/ingress-nginx/blob/mast ...
+
+#### GCE Ingress Controller: Does Not Support External, Dual-Stack Operation
+It appears that [GCE ingress controller](https://github.com/kubernetes/ingress-gce/blob/master/README.md#glbc) support is limited to IPv4-only access (QUESTION: Is this true???) from outside of the cluster. Therefore, GCE ingress controllers will not be capable of supporting external, dual-stack access to services running in a dual-stack Kubernetes cluster.
```
no, we allow V6, but I don't know if we allow both at the same time?  Knowing the data model, I expect we do.

pmichali commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+This proposal aims to extend the Kubernetes Pod Status API so that Kubernetes can track and make use of multiple IPv4/IPv6 address assignments per pod.
+
+#### Versioned API Change: PodStatus v1 core
+In order to maintain backwards compatibility for the core V1 API, this proposal adds a new "ExtraPodIPs" field to the core V1 version of the [PodStatus V1 core API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#podstatus-v1-core), while maintaining the existing "PodIP" field. This change is in accordance with the [Kubernetes API change
```
I guess the naive question is... Is the KEP supposed to be a functional spec (what), design spec (how), or some combination?

pmichali replied:
Makes sense.

lavalamp commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
````diff
@@ -0,0 +1,362 @@
+#### Internal Representation:
+The internal representation for the service IP (a.k.a. cluster IP) will be changed from a singular string to a slice of strings, and renamed "ClusterIPs":
+```
+    ClusterIPs []string `json:"clusterIPs,omitempty" protobuf:"bytes,3,opt,name=clusterIPs"`
````
I think you can use this construct in the external struct, too. Just require that ClusterIP == ClusterIPs[0]. Convert and default in a few places, and this is not a terrible user experience.

lavalamp replied:
Ah, I see there's already discussion about that for pod IPs--if cluster IP is already mutable then what I said probably won't work.

lavalamp commented:
re: feature branches... I don't know. If you're careful about syncing from master back into your branch, maybe it will be OK. I don't think this would conflict too much with the apply branch, so it might be worth thinking about. I wouldn't do two branches that make extensive changes like this in the same areas at the same time.

You currently would need a few people working on this that are on the admin list, as it's a bit spotty whether the merge bot will get around to feature branch PRs.

pmichali commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+- Cross-family connectivity: IPv4-to-IPv6 and IPv6-to-IPv4 connectivity is considered outside of the scope of this proposal, except in the case of ingress access (depending on ingress controller used).
+- CNI network plugins other than the Bridge and Host-Local IPAM plugins should support Kubernetes dual stack, but the development and testing of dual stack support for these other plugins is considered outside of the scope of this proposal.
+- Code changes should support any number IPv4 addresses and any number of IPv6 addresses per pod (up to some reasonable limit). However, testing will be done only up to the following limits:
+  - Pod addresses: 1 IPv4 address and 2 IPv6 addresses per pod maximum
```
^^ More reasons why it may be useful to call out use cases for dual-stack. One to identify the cases that could be addressed. Another is to see if the use cases **should** be addressed. And a third, to ensure we have an common understanding of how dual-stack is expected to be used.

leblancd replied:
@thockin - I'll switch this to 1 IPv4 + 1 IPv6 for this KEP. The support for multiple IPv6 addresses seems to be generating a lot of questions, and I don't find the 1 ULA + 1 GUA that compelling of a user case. If someone comes up with a compelling use case, we can change the test limits.

pmichali commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+- Kubernetes needs to be configurable for up to two service CIDRs.
+- Backend pods for a service can be dual stack. For the first release of dual-stack support, each IPv4/IPv6 address of a backend pod will be treated as a separate Kubernetes endpoint.
+- Kube-proxy needs to support IPv4 and IPv6 services in parallel (e.g. drive iptables and ip6tables in parallel).
+- Health/liveness/readiness probes should work for dual-stack pods. For the first release of dual-stack support, no additional configuration "knobs" will be added for probe definitions, and a probe is deemed successful if either an IPv4 or IPv6 response is received
```
That''s not what I'm saying, IIUC your response. I'm thinking that the goal is to see if the pod is working irrespective of the protocol used to communicate with the pod (of which you agree). As such, instead of having knobs to give the user the ability (and complexity) to select what protocol(s) to use, could we just pick a protocol. If only one protocol available, it is the same as today. If both available, pick one, with my preference being IPv6, as that is the end goal (IPv6 only) IMHO.

Later in this doc(L287-290) it talks about having config for both V4 and V6 health checks. I'm thinking we could just keep the original config with single address specified.

leblancd replied:
Here's a corner case to consider. I'm not sure how common this would be, but it may help clarify the tradeoffs we're talking about: Let's say there's a legacy IPv4-only application (one of the reasons we're adding dual stack) that's being run in an otherwise dual-stack cluster. One example is the  standard nginx dockerhub image. If you load this image in an (otherwise) dual-stack cluster, its pod would have both IPv4 and IPv6 addresses assigned, but the application, by default, will only bind to (listen on) IPv4 (bind address of 0.0.0.0, or all v4 IPs). If someone were to try to configure a health probe for this, it would fail if health probes only test for health on IPv6. On the other hand, if we design it so that we're trying health probes on both IPv4 and IPv6 (declaring health if either V4 OR V6 response is received), then the probes would work correctly with out-of-the-box nginx. If this scenario (and others like it) are not a big concern, then limiting probes to V6 only support for dual stack might be a worthwhile tradeoff.

As for kube-proxy's healthz bind address config (L287-290), I don't want to conflate this kube-proxy startup config with the per-probe-definition config. Sure, these are both related to health probes (for kube-proxy, anyway), but the big difference is that the per-probe-definition config knobs are tricky to do in a portable and non-controversial manner, whereas adding v4 and v6 bind address startup configs are fairly trivial, *if* we need them (the *if* being predicated on the tradeoff described above).

pmichali commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+- Kube-proxy needs to support IPv4 and IPv6 services in parallel (e.g. drive iptables and ip6tables in parallel).
+- Health/liveness/readiness probes should work for dual-stack pods. For the first release of dual-stack support, no additional configuration "knobs" will be added for probe definitions, and a probe is deemed successful if either an IPv4 or IPv6 response is received
+- Kubectl commands and output displays will need to be modified for dual-stack.
+- Kubeadm support will need to be added to enable spin-up of dual-stack clusters.
```
I think it's fine to select an orchestrator for the reference implementation. My question is whether the KubeAdm changes should be part of this effort or a companion effort. My vote is to keep the scope of an already large effort as small as possible.

leblancd replied:
Can we develop dual stack, along with CI, without Kubeadm changes?

pmichali replied:
I'm not saying that orchestrator changes are not needed. I'm suggesting breaking up the work into separate efforts ("a companion effort"). Are we going to scope out kubeadm changes in detail as part of this KEP, then?

leblancd replied:
If it's a required part of implementing and testing the dual stack feature, then I'm proposing that we need to mention that here. Specifically, I want to avoid saying that the kubeadm changes are out-of-scope, out of concern that the decoupled kubeadm dual-stack changes might get a lower priority relative to the rest of the dual-stack work. This would be in the spirit of a KEP intended to be a "feature, and effort tracking document" ([from KEP summary](https://github.com/kubernetes/community/blob/master/keps/0001-kubernetes-enhancement-proposal-process.md#summary)). For kubeadm implementation details, we can always file an issue against kubeadm,

pmichali commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+Older servers and clients that were built before the introduction of full dual stack will only be aware of and make use of the original, singular "PodIP" field. It is therefore considered to be the default IP address for the pod. This choice of which pod IP to choose as the default IP address is discussed in the following subsection.
+
+##### Default Pod IP Selection
+When the internal representation of the PodStatus (see the next section) is converted to the "v1 core" versioned API object (see the [Operational Overview section](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#operational-overview) of the API changes guidelines), a choice needs to be made as to which IP address from the "PodIPs" slice needs to be populated  ...
```
Can ignore, since I now understand the KEP needs to address design.

baodongli commented:
I understand that the goal of this KEP is to support dual stack on the POD network, and to make the IPs configured on the POD network available through k8s APIs. I noticed in the comments that it leans toward supporting one address per family only although the proposal suggests to support arbitrary number of IP addresses. However, the semantics of those addresses are not specified outside of dual stack context. If one address per family is supported, I think that there is an alternative approach to achieve the same without requiring API change. The idea is to use ipv4-embedded ipv6 addresses as specified in here:https://tools.ietf.org/html/rfc6052. This is how it might work:
       -- configure k8s cluster with both IPv6 CIDR and IPv4 CIDR
       -- configure k8s cluster with an IPv6 service CIDR (and optionally an IPv4 service CIDR, I see a lot of discussions around this in this KEP).
       -- cni-plugin's IPAM creates an ipv4-embedded ipv6 address based on the two CIDRs, and configure the pods to operate in dual stack mode.
       -- the ipv4-embedded ipv6 address is recorded in the podIP field of the POD status struct.
       -- K8s Services (such as kube-proxy) interprets the address based on if both service CIDRs or only one of the service CIDRs are configured.
       -- kubectl displays the podIPs based on if both CIDRs or only one of the C/IDRs are configured.

This would require much less change, test efforts, and can make the feature available with less effort and time.

idealhack commented:
/ok-to-test

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,362 @@
+- Service Access: IPv4-to-IPv4 and IPv6-to-IPv6 access from pods to Kubernetes services
+- External Server Access: IPv4-to-IPv4 and IPv6-to-IPv6 access from pods to external servers
+- Ingress Access: Access from IPv4 and/or IPv6 clients to Kubernetes services. Depending on the ingress controller used, connections may cross IP families (IPv4 clients to IPv6 services, or vice versa).
+- Service Discovery: Support for both A and AAAA records for kube-DNS and coreDNS
```
I think we should NOT plan on a kube-dns update.

@bowei or @johnbelamaric can argue with me now...

warmchang commented:
/retest

bgrant0607 commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,438 @@
+
+## Motivation
+
+The adoption of IPv6 has increased in recent years, and customers are requesting IPv6 support in Kubernetes clusters. To this end, the support of IPv6-only clusters was added as an alpha feature in Kubernetes Version 1.9. Clusters can now be run in either IPv4-only, IPv6-only, or in a very limited dual-stack configuration. This limited dual-stack support might be called "headless", in the sense t ...
```
Let's please not use the term "headless" here. It already has a specific meaning in the context of Kubernetes, which is no assigned service IP.

bgrant0607 commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,438 @@
+
+- Kubernetes needs to be made aware of multiple IPs per pod (limited to one IPv4 and one IPv6 address per pod maximum).
+- Link Local Addresses (LLAs) on a pod will remain implicit (Kubernetes will not display nor track these addresses).
+- Backend pods for a service can be dual stack. For the first release of dual-stack support, each IPv4/IPv6 address of a backend pod will be treated as a separate Kubernetes endpoint.
```
I do not think that duplicate pod endpoints is going to be acceptable. It affects more than just load balancing and DNS. Distributed applications use them for peer discovery. Monitoring systems such as Prometheus use them to identify monitoring targets. Users use them to determine how many instances are up.

leblancd replied:
If this is the case, then the [Endpoints V1 core API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpoints-v1-core) will need to change, specifically for the Subsets. I need to think through the details, but this would have the same backwards compatibility concerns as podIP/podIPs. The newer, or internal list of Subsets would have to couple each IPv4 address with its partner IPv6 (i.e. a V4/V6 pair from a given pod). This would then require all consumers (clients) of the Endpoints API to eventually use the dual-stack information... that's going to take some time.

feiskyer replied:
+1 for @bgrant0607's comments.

Could we change endpoints API to support multiple addresses for each subset?

leblancd replied:
Yes, I'll rewrite this to add support for multiple addresses per subset.

bgrant0607 commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,438 @@
+    nodePort: 30302
+    targetPort: 80
+    protocol: TCP
+    ipFamily: ipv4
```
I expect to eventually have to match other address/network properties here.

leblancd replied:
Can you clarify what you mean here?

bgrant0607 commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,438 @@
+
+### CoreDNS Operation
+
+Because this proposal retains support for only a single Kubernetes service CIDR, CoreDNS support should work with very little change, if any.
```
Headless services publish records for all ready endpoints.

thockin commented:
I am on mobile, so reading the whole KEP over is hard.  Is this assuming
that we have dual-stack Services, as opposed to the simpler single-stack
services,dual-stack pods?

On Wed, Jul 18, 2018, 1:38 PM Dane LeBlanc <notifications@github.com> wrote:

> *@leblancd* commented on this pull request.
> ------------------------------
>
> In keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md
> <https://github.com/kubernetes/community/pull/2254#discussion_r203521390>:
>
> > +- Multiple IPs vs. Dual-Stack: Code changes will be done in a way to facilitate future expansion to more general multiple-IPs-per-pod and multiple-IPs-per-node support. However, this initial release will impose "dual-stack-centric" IP address limits as follows:
> +  - Pod addresses: 1 IPv4 address and 1 IPv6 addresses per pod maximum
> +  - Node addresses: 1 IPv4 address and 1 IPv6 addresses per pod maximum
> +  - Service addresses: 1 IP address pod service
> +- For simplicity, the [Kubernetes Endpoints API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpoints-v1-core) will not be modified. Each IPv4/IPv6 address that is assigned to a backend pod will be treated and displayed as a separate endpoint for the corresponding service. Service load balancing (e.g. using kube-proxy) will be performed on a per-address basis, rather than being done on a per-dual-stack-endpoint basis.
> +- Kube-DNS is expected to be End-of-Life soon, so dual-stack testing will be performed using coreDNS.
> +- External load balancers that rely on Kubernetes services for load balancing functionality will only work with the IP family that matches the IP family of the cluster's service CIDR.
> +- Dual-stack support for Kubernetes orchestration tools other than kubeadm (e.g. miniKube, KubeSpray, etc.) are considered outside of the scope of this proposal.
> +
> +## Proposal
> +
> +In order to support dual-stack in Kubernetes clusters, Kubernetes needs to have awareness of and support dual-stack addresses for pods and nodes. Here is a summary of the proposal (details follow in subsequent sections):
> +
> +- Kubernetes needs to be made aware of multiple IPs per pod (limited to one IPv4 and one IPv6 address per pod maximum).
> +- Link Local Addresses (LLAs) on a pod will remain implicit (Kubernetes will not display nor track these addresses).
> +- Backend pods for a service can be dual stack. For the first release of dual-stack support, each IPv4/IPv6 address of a backend pod will be treated as a separate Kubernetes endpoint.
>
> If this is the case, then the Endpoints V1 core API
> <https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpoints-v1-core>
> will need to change, specifically for the Subsets. I need to think through
> the details, but this would have the same backwards compatibility concerns
> as podIP/podIPs. The newer, or internal list of Subsets would have to
> couple each IPv4 address with its partner IPv6 (i.e. a V4/V6 pair from a
> given pod). This would then require all consumers (clients) of the
> Endpoints API to eventually use the dual-stack information... that's going
> to take some time.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/2254#discussion_r203521390>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVN2Jc6qUnQxO2dGhSM6qtVOGCT2Qks5uH5zdgaJpZM4Ukyr8>
> .
>

leblancd commented:
@thockin - I think in either case (dual-stack services or even single-stack services), we'd still want dual-stack endpoints. The reason being that we'd want listeners/consumers of endpoints to have dual-stack access to pods. This would help with providing dual-stack access from the outside (even when services are single-family), e.g. with ingress controllers or external load balancers that map directly to endpoints rather than mapping to service IPs.

pmichali commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,438 @@
+- Kube-proxy needs to drive iptables and ip6tables in parallel. This is required in order to allow exposing services via both IPv4 and IPv6, e.g. using Kubernetes:
+  - NodePort
+  - ExternalIPs
+- "Nice to Have" or stretch goal: IPVS proxying needs to support IPv4 and IPv6 services in parallel. However, this support depends upon resolution of [cloudnativelabs/kube-router Issue #307](https://github.com/cloudnativelabs/kube-router/issues/307).
```
May want to split this into two items. One is IPVS via kube-router (as a replacement for kube-proxy). That depends on issue 307 for IPv6 support. The other is kube-proxy running in IPVS mode, which should work, provided the mentioned kube-proxy changes in this KEP are performed.

leblancd replied:
@pmichali - Do you know which IPVS mode is (eventually) going to be the reference mode, IPVS via kube-router or IPVS via kube-proxy?

pmichali replied:
I don't know which method is planned for the reference in the future. Maybe someone in the community has a better handle on that.

For kube-proxy, it appears that the default is still to use iptables currently.  I think kube-router is pushing for GA, but I don't know the timing or the plans for either of these two.

I have to admit, kube-router is a much easier way to set things up.

feiskyer replied:
I think we should focus on ipvs mode here, because it's part of core k/k

leblancd replied:
@feiskyer - Sounds good, especially if you can help us out with getting IPVS mode working (V6-only first, then eventually dual-stack).

rkamudhan commented:
Is there any PR of code on Kube-proxy, we can look into ?

leblancd commented:
@rkamudhan - No kube-proxy related PRs yet. We've been focusing on the design spec and CI. The kube-proxy related changes shouldn't be too complicated, since it will only need to support NodePort and ExternalIPs in the iptables for the non-preferred IP family.

pmichali commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,438 @@
+
+In the future, we may want to consider adding a "dual-stack" option for the "--preferred-probe-ip-family" argument, indicating that a kubelet should test for probes using both IPv4 and IPv6 addresses for a pod, and consider the probe successful if a response is received via either address.
+
+### CoreDNS Operation
```
Maybe it is a provisioning question, but, what are the plans for supporting the DNS server in dual-stack mode? Provide both an IPv4 and IPv6 address to pods? There are some things, like the the bind9 container, where the DNS server IP must be provided. Wondering how we handle that in dual-stack mode.

leblancd replied:
Pod access to the CoreDNS server will be done via the CoreDNS service IP, which will be single-family. As for DNS records that CoreDNS will return to the pods, the records should be single-family for non-headless services, and potentially dual-stack (both A and AAAA) for headless services.

asauber commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,531 @@
+
+### Non-Goals
+
+- Service CIDRs: Dual-stack service CIDRs will not be supported for this proposal. Service access within a cluster will be done via all IPv4 service IPs or all IPv6 service IPs.
```
I'm facing an issue in which dual-stack NodePort Services specifically would be very useful. I need to provision LoadBalancers with IPv4 backends that are Nodes in an IPv6 cluster. 

Can we specify ClusterIP here specifically and propose dual-stack NodePorts?

leblancd replied:
@asauber - I had included the dual-stack NodePort support (and dual-stack externalIPs support) in the Kube-Proxy section, but I'll add it to the Goals section to make it clearer.

leblancd commented:
@thockin - Could you take a look at this when you get a chance? There have been several changes since the last time you reviewed that deserve some thought/discussion, e.g.:
- Single-family service CIDR
- Dual-stack endpoint API change
- "endpointFamily" configuration for services
I've also opened 2 API review PRs in architecture-tracking (one for PodStatus, one for Endpoint API change), and there are a couple of coding PRs (PR #69029 and PR #69153).
Thanks!

bowei commented:
cc: @varunmar

rkamudhan commented:
@leblancd Regarding dual stack endpoints, Is that each endpoints subset have both IPv4 and IPv6 address ?

leblancd commented:
@rkamudhan - Yes, each endpoints subset will have both IPv4 and IPv6 addresses. Our initial proposal was to treat each IP address as separate endpoints (IPv4 address was one endpoint, IPv6 address would be a different endpoint, so 2 endpoints per backend). However, we were advised that the 2 endpoints per backend was problematic for the reasons listed here: https://github.com/kubernetes/community/pull/2254/files#diff-bd8f8bd5daa9561082402a2ff4795f05R279
The current proposal is to modify the "EndpointAddress" structure in the API to include an IPs[] list (see https://github.com/kubernetes/community/pull/2254/files#diff-bd8f8bd5daa9561082402a2ff4795f05R293).

leblancd commented:
@sb1975 - The target release for Kubernetes dual stack is now Release 1.14. The schedule for release 1.14 has not been finalized (release schedules are currently published only up to 1.13, see https://github.com/kubernetes/sig-release/tree/master/releases.
Setting up IPv4 ingress in parallel with IPv6 ingress with current Kubernetes *might* be possible, but complex. e.g. you could try setting up an IPv6-only cluster, and then add stateless NAT46 at the cluster edge for access from IPv4 external clients to IPv6 Kubernetes services. It might also be possible to use an NGINX ingress controller that has dual-stack access from external, but maps to single-family Kubernetes nodePorts. I haven't tried either approach, so YMMV.

sb1975 commented:
@leblancd - Thanks for the details, do you have any link to understand the steps involved in creating the multiple ingress- for IPv4 and IPv6 seperately and also how to configure the NAT46 in the IPv4 ingress.

leblancd commented:
@sb1975 - With the help of @aojea, we've put together an overview on how to install a dual-stack NGINX ingress controller on an (internally) IPv6-only cluster: ["Installing a Dual-Stack Ingress Controller on an IPv6-Only Kubernetes Cluster](https://github.com/leblancd/kube-v6/tree/master/dual-stack-ingress). This requires that the nodes be configured with dual-stack public/global IPv4/IPv6 addresses, and it runs the ingress controller pods on the host network of each node.

I haven't configured Stateless NAT46 on a Kubernetes IPv6-only cluster, but you can find some good background references on the web. e.g. Citrux has a helpful reference for configuring their NAT46 appliance [here](https://docs.citrix.com/en-us/netscaler/12/networking/ip-addressing/configuring-network-address-translation/stateless-nat46-translation.html), and there's a video on configuring Stateless NAT46 on a  Cisco ASA [here](http://www.labminutes.com/sec0104_asa_9_nat46_nat64_twice_nat_1).

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+       kube-master# kubectl get endpoints
+       NAME            ENDPOINTS                             AGE
+       kubernetes      {[fd00::100]:6443,10.0.0.2:6643}      15m
+       nginx-service   {[fd00:db8:1::2]:80,192.168.1.3:80}   20m
```
Do we have precedent for multi-line output like this?  @kubernetes/sig-cli-api-reviews

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
````diff
@@ -0,0 +1,626 @@
+```
+  --bind-address  stringSlice   (IP addresses, in a comma separated list, Default: [0.0.0.0,])
+```
+Only the first address of each IP family will be used; all others will be ignored.
````
As above, any reason not to throw an error rather than ignore?

leblancd replied:
Yes, better to flag this as bad startup configuration, and keep the expected behavior clear.

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
````diff
@@ -0,0 +1,626 @@
+```
+  --cluster-cidr  ipNetSlice   (IP CIDRs, in a comma separated list, Default: [])
+```
+Only the first CIDR for each IP family will be used; all others will be ignored.
````
ditto

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+
+### IPVS Support and Operation
+
+Since IPVS functionality does not yet include IPv6 support (see [cloudnativelabs/kube-router Issue #307](https://github.com/cloudnativelabs/kube-router/issues/307)), support for IPVS functionality in a dual-stack cluster is considered a "nice-to-have" or stretch goal.
```
I didn't realize teh IPVS kube-proxy doesn't do v6 at all - is that right?

@m1093782566 can you comment on this?  I am reluctant to GA without it

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
````diff
@@ -0,0 +1,626 @@
+
+Currently, health, liveness, and readiness probes are defined without any concern for IP addresses or families. For the first release of dual-stack support, a cluster administrator will be able to select the preferred IP family to use for probes when a pod has both IPv4 and IPv6 addresses. For this selection, a new "--preferred-probe-ip-family" argument for the for the [kubelet startup configurat ...
+```
+  --preferred-probe-ip-family  string   ["ipv4", "ipv6", or "none". Default: "none", meaning use the pod's default IP]
````
Why is this a global flag rather than per-pod or per-probe?

leblancd replied:
See my response to your earlier comment. I think we can do without this configuration, and for probes Kubelet should use the family of the default IP for each pod. I don't think we need a per-pod or per-probe configuration in the initial release of dual-stack, maybe to this as a followup (including support of probes that work on both IP families, requiring either both V4 and V6 responses, or either V4 or V6 responses).

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+
+- Because service IPs will remain single-family, pods will continue to access the CoreDNS server via a single service IP. In other words, the nameserver entries in a pod's /etc/resolv.conf will typically be a single IPv4 or single IPv6 address, depending upon the IP family of the cluster's service CIDR.
+- Non-headless Kubernetes services: CoreDNS will resolve these services to either an IPv4 entry (A record) or an IPv6 entry (AAAA record), depending upon the IP family of the cluster's service CIDR.
+- Headless Kubernetes services: CoreDNS will resolve these services to either an IPv4 entry (A record), an IPv6 entry (AAAA record), or both, depending on the service's endpointFamily configuration (see [Configuration of Endpoint IP Family in Service Definitions](#configuration-of-endpoint-ip-family-in-service-definitions)).
```
Depends on previous question about this config being per-service vs per-pod

leblancd replied:
I now think single-family headless services would work (on par with non-headless kubernetes services being single-family).

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+The [Kubernetes ingress feature](https://kubernetes.io/docs/concepts/services-networking/ingress/) relies on the use of an ingress controller. The two "reference" ingress controllers that are considered here are the [GCE ingress controller](https://github.com/kubernetes/ingress-gce/blob/master/README.md#glbc) and the [NGINX ingress controller](https://github.com/kubernetes/ingress-nginx/blob/mast ...
+
+#### GCE Ingress Controller: Out-of-Scope, Testing Deferred For Now
+It is not clear whether the [GCE ingress controller](https://github.com/kubernetes/ingress-gce/blob/master/README.md#glbc) supports external, dual-stack access. Testing of dual-stack access to Kubernetes services via a GCE ingress controller is considered out-of-scope until after the initial implementation of dual-stack support for Kubernetes.
```
I'd say this is Google's problem to implement.

@bowei

leblancd replied:
I'll just say this is out-of-scope for this effort.

varunmar replied:
The unclear parts should at least be clarified :)
I can take this one.

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+
+The ClusterIP service type will be single stack, so for this case there will be no changes to the current load balancer config. The user has the option to create two load balancer IP resources, one for IPv6 and the other for IPv4, and associate both with the same application instances.
+
+#### Type NodePort
```
Curious - how important is this requirement?  What if NodePorts were single-stack, at least initially?  Woudl it make this more tractable?

leblancd replied:
I think that support of dual-stack NodePorts is an important requirement, specifically for providing dual-stack access from outside of the cluster. For example, if you look at the way that the NGINX ingress controller can be configured on bare metal (see [Bare-metal considerations](https://kubernetes.github.io/ingress-nginx/deploy/baremetal/), there are several installation options (MetalLB, over NodePort, via host network, self-provisioned edge, over externalIPs, etc.), but all of these options except the host network option require dual-stack nodePorts or externalIPs.

Supporting dual-stack nodePorts will require some code refactoring (or change in workflow) in kube-proxy (i.e. should be able to set up NodePort in iptables without requiring simultaneous configuration of service IPs), but I don't think that will be major work.

We could consider phasing this in later, but IMO dual-stack Kubernetes support isn't useful without dual-stack nodePorts and externalIPs.

thockin replied:
Sadly, that is what I thought you'd say, which means we NEED dual-stack endpoints, right?

leblancd replied:
Yes, I believe so.

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+
+#### Multiple bind addresses configuration
+
+The existing "--bind-address" option for the will be modified to support multiple IP addresses in a comma-separated list (rather than a single IP string).
```
Why is this a sub-heading of cloud-providers?

There are other components that support a flag like this - do we have a list?

leblancd replied:
@rpothier ?

rpothier replied:
Kube-proxy and the kubelet startup config have a similar requirement, that's a good idea to list them.
Also possibly the controller manager if we went with the full Dual Stack.

rpothier replied:
Also, that line is missing the link to the cloud controller manager, so it should read
The existing "--bind-address" option for the [cloud-controller-manager](https://kubernetes.io/docs/reference/command-line-tools-reference/cloud-controller-manager/) will be modified ...

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+  - name: MY_POD_IPS
+    valueFrom:
+      fieldRef:
+        fieldPath: status.podIPs
```
@kubernetes/sig-cli-api-reviews We should get a consult as to whether this is right or whether the fieldpath should be something like `status.podIPs[].ip` - it was supposed to be a literal syntax.

squeed commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+    //    Properties: Arbitrary metadata associated with the allocated IP.
+    type PodIPInfo struct {
+        IP string
+        Properties map[string]string
```
The properties map would be very useful for multi-network. And no sense in changing the data structure twice? I'd prefer to keep it if possible.

varunmar replied:
I see this has already generated some discussion :)
I'll add my own comments here - @leblancd you can ignore my other comment up above.

One pattern I've seen used successfully in internal interfaces is to have a mostly-strongly typed struct with a bag-of-strings at the end for "experimental" free-for-all properties. But that requires agreement that the bag-of-strings should not be relied upon, and can change at any time. I do not think we can enforce such a thing if we put this into an external facing API, so my vote is to only add fully-typed fields with validation and strong semantic meaning. Then we can argue about names all at once before they are used instead of after the fact :)

squeed commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+This internal representation should eventually become part of a versioned API (after a period of deprecation for the singular "PodIP" field).
+
+#### Maintaining Compatible Interworking between Old and New Clients
+Any Kubernetes API change needs to consider consistent interworking between a possible mix of clients that are running old vs. new versions of the API. In this particular case, however, there is only ever one writer of the PodStatus object, and it is the API server itself. Therefore, the API server does not have an absolute requirement to implement any safeguards and/or fixups between the singula ...
```
Oh, I would expect someone to start writing multiple podIPs faster than you could imagine.

squeed commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
````diff
@@ -0,0 +1,626 @@
+    targetPort: 9376
+  endpointFamily: ipv4
+```
+This per-service endpoint IP family configuration is required for IPv4-only and IPv6-only Application Support. For example, if there is a legacy application that only binds to IPv4 addresses (IPv4-only application) that is running in an otherwise dual-stack cluster, then we don't want endpoints for the service to include IPv6 addresses that are allocated to backend pods for the service. Otherwise ...
````
Either way, this is useful information for a service (and pod ports, while we're at it), because there's nothing stopping a program from binding some things as v4 only and others as v6 only. A bad idea, sure, but technically plausible.

thockin replied:
I don't want to design to hypotheticals, especially API surface.  If I understand, the worst case is that we publish a v6 nodeport with v6 endpoints and there's nothing listening there.  Is that right?

If so, that doesn't seem like the end of the world to me - I'd rather defer this and come back to it if it is a real problem.

squeed commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+This feature requires the use of the [CNI Networking Plugin API version 0.3.1](https://github.com/containernetworking/cni/blob/spec-v0.3.1/SPEC.md)
+or later. The dual-stack feature requires no changes to this API.
+
+The versions of CNI plugin binaries that must be used for proper dual-stack functionality (and IPv6 functionality in general) depend upon the version of Docker that is used in the cluster nodes (see [CNI issue #531](https://github.com/containernetworking/cni/issues/531) and [CNI plugins PR #113](https://github.com/containernetworking/plugins/pull/113)):
```
Issue 531 was a weird docker interaction that we've fixed; we no longer have a Docker version dependency.

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+- Link Local Addresses (LLAs) on a pod will remain implicit (Kubernetes will not display nor track these addresses).
+- For simplicity, only a single family of service IPs per cluster will be supported (i.e. service IPs are either all IPv4 or all IPv6).
+- Backend pods for a service can be dual stack.
+- Endpoints for a dual-stack backend pod will be represented as a dual-stack address pair (i.e. 1 IPv4/IPv6 endpoint per backend pod, rather than 2 single-family endpoints per backend pod)
```
I think the non-goal is correct -- Kubernetes itself is NOT doing address family translation.  The fact that Ingress controllers can do that is merely a side-effect of the fact that they are almost universally full proxies which accept the frontside connection and open another for the backside.  I don't think it is obvious that we won't convert a v4 to a v6 connection via service VIP, and we should be clear that it is NOT part of the scope.

When I wrote this question I was thinking about Service VIP -> Backends.  That has to be same-family (because of the non-goal above), so alt-family endpoints is nonsensical.  I think this is also true for external IPs and load-balancer IPs -- they are received and processed the same as cluster IPs, with no family conversion.

BUT...

1) external IPs is a list, so could plausibly include both families.
2) LB status.ingress is a list, so could plausibly include both families.
3) nodePorts could reasonably be expected to work on any interface address.
4) headless services could reasonably be expected to work on any interface address.

You suggest you might be willing to step back from (4), but unless we also step back from (1), (2), AND (3), that wouldn't save any work.

I think the only simplification here is if we can say "all service-related logic is single family".  And I am not sure that is very useful -- tell me if you disagree.

Assuming we have ANY service-related functionality with both families, we need dualstack endpoints :(

Or am I missing something?

thockin replied:
...more to the point, is multi-family endpoints, nodeports, LBs, etc somethign we can defer to a "phase 2" and iterate?  Would it simplify this proposal or just make it not useful?

thockin replied:
OK, so any scope-reduction around not doing dual-stack endpoints is rendered moot,a nd all such comments should be ignored :)

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+        Properties map[string]string
+    }
+
+    // IP addresses allocated to the pod with associated metadata. This list
```
I'd spell it out in the comments.  I don't expect end-users to read our API devel docs :)

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+
+However, as a defensive coding measure and for future-proofing, the following API version translation logic will be implemented for the PodIP/PodIPs fields:
+
+##### V1 to Core (Internal) Conversion
```
for this doc, pointing to the API docs is OK.  For end-user documentation, we should spell out the behavior.

thockin commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+     // The default IP for this endpoint.
+     IP string `json:"ip" protobuf:"bytes,1,opt,name=ip"`
+     // The IPs for this endpoint. The zeroth element (IPs[0] must match
+     // the default value set in the IP field)
```
I think it's all or nothing - I am afraid I talked myself back into needing dual-stack endpoints, unless we can drop all dual-stack services-related work and iterate on that

sb1975 commented:
> @sb1975 - With the help of @aojea, we've put together an overview on how to install a dual-stack NGINX ingress controller on an (internally) IPv6-only cluster: ["Installing a Dual-Stack Ingress Controller on an IPv6-Only Kubernetes Cluster"](https://github.com/leblancd/kube-v6/tree/master/dual-stack-ingress). This requires that the nodes be configured with dual-stack public/global IPv4/IPv6 addresses, and it runs the ingress controller pods on the host network of each node.
> 
> I haven't configured Stateless NAT46 on a Kubernetes IPv6-only cluster, but you can find some good background references on the web. e.g. Citrux has a helpful reference for configuring their NAT46 appliance [here](https://docs.citrix.com/en-us/netscaler/12/networking/ip-addressing/configuring-network-address-translation/stateless-nat46-translation.html), and there's a video on configuring Stateless NAT46 on a Cisco ASA [here](http://www.labminutes.com/sec0104_asa_9_nat46_nat64_twice_nat_1).

@leblancd : Thanks for the response , this is very helpful but we have a additional use case : we need a IPv4 client reach a Kubernetes IPv6 service over non-http traffic(like SNMP). Now I understand the ingress would only support only http rules so how do we enable this please ?

rpothier commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
````diff
@@ -0,0 +1,626 @@
+The kubeadm configuration options for advertiseAddress and podSubnet will need to be changed to handle a comma-separated list of CIDRs:
+```
+    api:
+      advertiseAddress: "fd00:90::2,10.90.0.2" [Multiple IP CIDRs, comma separated list of CIDRs]
````
Nit: the advertiseAddresses are addresses, not CIDRs

aojea commented:
> @leblancd : Thanks for the response , this is very helpful but we have a additional use case : we need a IPv4 client reach a Kubernetes IPv6 service over non-http traffic(like SNMP). Now I understand the ingress would only support only http rules so how do we enable this please ?

the ingress controller supports non http traffic over TCP and UDP, however, seems that feature is going to be removed from the nginx ingress controller https://github.com/kubernetes/ingress-nginx/pull/3197

thockin commented:
@leblancd Can you make your edits and resolve any comment threads that are old and stale?  Ping me and I'll do another top-to-bottom pass.  Hopefully we can merge soon and iterate finer points.

leblancd commented:
@thockin - Will get the next editing pass in by Monday. I'm working on some V6 CI changes today. Thanks!

neolit123 commented:
/assign @timothysc 
@kubernetes/sig-cluster-lifecycle 
for the kubeadm and kubespray related topics.

varunmar commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+## Motivation
+
+The adoption of IPv6 has increased in recent years, and customers are requesting IPv6 support in Kubernetes clusters. To this end, the support of IPv6-only clusters was added as an alpha feature in Kubernetes Version 1.9. Clusters can now be run in either IPv4-only, IPv6-only, or in a "single-pod-IP-aware" dual-stack configuration. This "single-pod-IP-aware" dual-stack support is limited by the f ...
+- Some CNI network plugins are capable of assigning dual-stack addresses on a pod, but Kubernetes is aware of only one address per pod.
```
Thanks for the detailed design! We are running some prototype dual stack configurations inside of GCE and starting to find ways to work around the fact that Kubernetes itself is unaware of the IPv6 addresses.
I'd like very much to stay close to the design in this doc, and reinforce our prototyping/testing efforts when the time comes. Please keep me in the loop :)

varunmar commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+This proposal aims to extend the Kubernetes Pod Status API so that Kubernetes can track and make use of up to one IPv4 address and up to one IPv6 address assignment per pod.
+
+#### Versioned API Change: PodStatus v1 core
+In order to maintain backwards compatibility for the core V1 API, this proposal retains the existing (singular) "PodIP" field in the core V1 version of the [PodStatus V1 core API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#podstatus-v1-core), and adds a new array of structures that store pod IPs along with associated metadata for that IP. The metadata for each IP (refer  ...
```
I've grown really leery of a bag-of-strings properties approach - I think that unless we commit to a naming scheme, or at least a conflict resolution mechanism if two components start using the same keys in incompatible ways I would really like to see these develop as fully specified types rather than a loose bag of strings.
What do you think?

varunmar commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
````diff
@@ -0,0 +1,626 @@
+```
+  --pod-cidr  ipNetSlice   [IP CIDRs, comma separated list of CIDRs, Default: []]
+```
+Only the first address of each IP family will be used; all others will be ignored.
````
Also, I think you mean "first CIDR" here, not the first address.

varunmar commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
```diff
@@ -0,0 +1,626 @@
+     IP string `json:"ip" protobuf:"bytes,1,opt,name=ip"`
+     // The IPs for this endpoint. The zeroth element (IPs[0] must match
+     // the default value set in the IP field)
+     IPs []string `json:"ips" protobuf:"bytes,5,opt,name=ips"`
```
If the pod IPs have metadata describing them (the PodIPs struct) then isnt' that useful to surface here as well?
It seems like @squeed has a use case for labeling IPs with the network names - is it useful to endpoints controllers (like ingress?) to know that here too?

If we do have the metadata here though, it will need to be exactly the same structure as the PodIPs. I'm not sure if that raises any other API issues.

varunmar commented on keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md:
````diff
@@ -0,0 +1,626 @@
+#### Configuration of Endpoint IP Family in Service Definitions
+This proposal adds an option to configure an endpoint IP family for a Kubernetes service:
+```
+    endpointFamily: <ipv4|ipv6|dual-stack>       [Default: dual-stack]
````
If service addresses only come from a single address family, why does this belong in the Service definition?

Or to put it another way - shouldn't the default be the same address family as the service CIDR? If Kubernetes itself isn't going to do any 6/4 translation, could you say more about how this can be used in any other way?

timothysc commented:
@neolit123 This will affect sig-cluster-lifecycle, but it's squarely on sig-networking. 
/assign @thockin

justaugustus commented:
REMINDER: KEPs are moving to [k/enhancements](https://github.com/kubernetes/enhancements) on November 30. Please attempt to merge this KEP before then to signal consensus.
For more details on this change, review [this thread](https://groups.google.com/d/msg/kubernetes-dev/LIkZoIqCT20/t515CV2mAgAJ).

Any questions regarding this move should be directed to that thread and not asked on GitHub.

justaugustus commented:
KEPs [have moved](https://github.com/kubernetes/enhancements/pull/638) to [k/enhancements](https://github.com/kubernetes/enhancements). 
This PR will be closed and any additional changes to this KEP should be submitted to k/enhancements.
For more details on this change, review [this thread](https://groups.google.com/d/msg/kubernetes-dev/LIkZoIqCT20/t515CV2mAgAJ).

Any questions regarding this move should be directed to that thread and not asked on GitHub.
/close

aojea commented:
Moving kubernetes/community#2254 to https://github.com/kubernetes/enhancements/pull/648
